File: backend/app/schemas/boq.py
from decimal import Decimal

from pydantic import BaseModel, Field


class BOQSectionCreate(BaseModel):
    name: str = Field(min_length=1)
    code: str | None = None
    description: str | None = None
    sort_order: int = Field(default=0, ge=0)


class BOQSectionResponse(BOQSectionCreate):
    id: str
    project_id: str


class BOQItemCreate(BaseModel):
    item_no: int = Field(ge=1)
    description: str = Field(min_length=1)
    unit: str = Field(min_length=1)
    quantity: Decimal = Field(gt=0)
    unit_rate: Decimal = Field(ge=0)
    rate_analysis_id: str | None = None
    section_id: str | None = None


class BOQItemResponse(BOQItemCreate):
    id: str
    amount: Decimal

File: frontend/src/api/boq.ts
import { apiFetch } from "./client";

export type BOQSection = {
  id: string;
  project_id: string;
  name: string;
  code?: string | null;
  description?: string | null;
  sort_order: number;
};

export type BOQItem = {
  id: string;
  item_no: number;
  description: string;
  unit: string;
  quantity: number;
  unit_rate: number;
  amount: number;
  section_id?: string | null;
  rate_analysis_id?: string | null;
};

export async function listSections(projectId: string): Promise<BOQSection[]> {
  return apiFetch<BOQSection[]>(`/projects/${projectId}/boq/sections`);
}

export async function createSection(
  projectId: string,
  payload: Partial<BOQSection>,
): Promise<BOQSection> {
  return apiFetch<BOQSection>(`/projects/${projectId}/boq/sections`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export async function listBOQ(projectId: string): Promise<BOQItem[]> {
  return apiFetch<BOQItem[]>(`/projects/${projectId}/boq`);
}

export async function createBOQ(
  projectId: string,
  payload: Partial<BOQItem>,
): Promise<BOQItem> {
  return apiFetch<BOQItem>(`/projects/${projectId}/boq`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}
